### DISCUSSION TITLE
os.exec()中无法调用vs prompt中才有的命令（fxc）

### DISCUSSION THREAD

**Discussion by JX-Master** [xmake-io/xmake#3169] [Q&A] [answered]
需求：我需要在构建应用程序的时候使用`fxc.exe`编译部分着色器代码。目前我是在应用程序的`before_build`函数中调用`os.exec("fxc ....")`实现的。程序在尝试调用时会提示找不到fxc可执行文件。经过排查，原因在于`os.exec()`使用的是和普通cmd/shell的环境，因此不包括类似`cl`、`fxc`等vs prompt中才有的命令。

我想知道是否有一种方法可以让`os.exec`的执行环境和vs prompt一致，以便我们能够正确调用这些只有在vs prompt中才能使用的命令。如果是需要设置查找路径，那么是否有一种方法可以获取当前Windows SDK的路径（类似`C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\%arch%`），以便设置程序的查找路径。目前尝试使用`detect.find_program("fxc")`会返回`nil`。

谢谢。

**Comment by waruqi** [comment, marked as answer]
```lua
function (target)
    local msvc = target:toolchain("msvc")
    local runenvs = msvc:runenvs()
    local cl = target:tool("cc")
    local fxc = cl:gsub("cl%.exe", "fxc.exe")
    os.execv(fxc, {}, {envs = runenvs})
end
```

**Reply by JX-Master** [reply]
cl和fxc不在一个路径里。在我的电脑上，cl的路径是`C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\cl.exe`，fxc的路径是`C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\x64\fxc.exe`，所以查找`C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\fxc.exe`会找不到。

**Reply by waruqi** [reply]
```lua
import("lib.detect.find_tool")
function (target)
    local msvc = target:toolchain("msvc")
    local runenvs = 
    local fxc = find_tool("fxc", {envs = runenvs})
    if fxc then
        os.execv(fxc.program, {}, {envs = runenvs})
    end
end
```

**Reply by JX-Master** [reply]
find_tool返回了nil，我不太确定envs参数在这里是否起作用了，然后fxc所在的路径在msvc:runenvs().PATH中有。

**Reply by waruqi** [reply]
默认只会尝试执行 fxc --version 来检测查找，如果不是，需要自己适配下，参考：https://github.com/xmake-io/xmake/blob/master/xmake/modules/detect/tools/find_cl.lua

自己改下脚本调下。。

**Reply by JX-Master** [reply]
我发现

`os.execv("fxc", {"/T", "vs_5_0", vs_source, "/Fo", vs_path, "/E", "main", "/O3"}, {envs = {PATH = "C:\\Program Files (x86)\\Windows Kits\\10\\bin\\10.0.19041.0\\x64"}})`

这个也无法调用查找到fxc，会提示No such file or directory，只有

`os.execv("C:\\Program Files (x86)\\Windows Kits\\10\\bin\\10.0.19041.0\\x64\\fxc", {"/T", "vs_5_0", vs_source, "/Fo", vs_path, "/E", "main", "/O3"})`

可以正常运行。按道理如果在envs里面提供PATH的话不应该会找不到，不知道是什么原因。

**Reply by waruqi** [reply]
> 按道理如果在envs里面提供PATH的话不应该会找不到，不知道是什么原因。

传 envs 给 os.execv 仅仅只是创建进程成功后，进程绑定到对应的 envs，然后进程内部环境切到了 envs，但前提是进程本身能够被找到。

传 envs 给 os.execv/CreateProcess 不能保证 fxc 本身能够被运行

所以需要 find_tool/find_program 里传envs，并适配对应的命令参数，才能找到完整路径的 fxc ，它会自动从 envs 的 PATH 里帮你找 fxc，前提上找到后，能够尝试执行成功，否则算检测失败。也就是 find_tool/find_program 干了 shell 里面从 PATH 找进程的事情。

**Reply by JX-Master** [reply]
解决了，最终的代码如下，供大家参考：
```
local msvc = target:toolchain("msvc")
local runenvs = msvc:runenvs()
local opt = {}
opt.envs = runenvs
opt.check = function(program) end -- 绕过--version检查，因为fxc没有实现该功能。
local fxc = find_program("fxc", opt)
os.execv(fxc, {...})
```
